File: reference/reference.net/db4o-reference-chapters/src/Db4objects.Db4odoc/QBE/QBEExample.cs
/* Copyright (C) 2007 db4objects Inc. http://www.db4o.com */
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;

using Db4objects.Db4o;
using Db4objects.Db4o.Ext;
using Db4objects.Db4o.Query;

namespace Db4objects.Db4odoc.QBE
{
    class QBEExample
    {
        private const string Db4oFileName = "reference.db4o";

        private const int ObjectCount = 10;

        private static IObjectContainer _container = null;

        public static void Main(string[] args)
        {
            File.Delete(Db4oFileName);
         
            Test();
            Test1();
            Test2();
            Test3();
            Test4();
        }
        // end Main

        private static IObjectContainer Database()
        {
            if (_container == null)
            {
                try
                {
                    _container = Db4oFactory.OpenFile(Db4oFileName);
                }
                catch (DatabaseFileLockedException ex)
                {
                    System.Console.WriteLine(ex.Message);
                }
            }
            return _container;
        }

        // end Database

        private static void CloseDatabase()
        {
            if (_container != null)
            {
                _container.Close();
                _container = null;
            }
        }

        // end CloseDatabase

        private static void Test()
        {
            IObjectContainer container = Database();
            if (container != null)
            {
                try
                {
                    Pilot pilot = new Pilot("Kimi Raikonnen", 100);
                    container.Set(pilot);
                    container.Commit();
                    IObjectSet result = container.Get(new Pilot("Kimi Raikonnen", 100));
                    ListResult(result);
                }
                catch (Db4oException ex)
                {
                    System.Console.WriteLine("Db4o Exception: " + ex.Message);
                }
                catch (Exception ex)
                {
                    System.Console.WriteLine("System Exception: " + ex.Message);
                }
                finally
                {
                    CloseDatabase();
                }
            }
        }

        // end Test

        private static void Test1()
        {
            IObjectContainer container = Database();
            if (container != null)
            {
                try
                {
                    // Pilot1 contains re-initialisation in the constructor
                    Pilot1 pilot = new Pilot1("Kimi Raikonnen", 100);
                    container.Set(pilot);
                    container.Commit();
                    // QBE returns result with wrong points
                    IObjectSet result = container.Get(new Pilot1("Kimi Raikonnen", 100));
                    System.Console.WriteLine("Test QBE on class with member re-initialization in constructor");
                    ListResult(result);
                }
                catch (Db4oException ex)
                {
                    System.Console.WriteLine("Db4o Exception: " + ex.Message);
                }
                catch (Exception ex)
                {
                    System.Console.WriteLine("System Exception: " + ex.Message);
                }
                finally
                {
                    CloseDatabase();
                }
            }
        }

        // end Test1

        private static void Test2()
        {
            IObjectContainer container = Database();
            if (container != null)
            {
                try
                {
                    // Pilot1Derived derives the constructor with re-initialisation
                    Pilot1Derived pilot = new Pilot1Derived("Kimi Raikonnen", 100);
                    container.Set(pilot);
                    container.Commit();
                    // QBE returns result with wrong points
                    IObjectSet result = container.Get(new Pilot1Derived("Kimi Raikonnen", 100));
                    System.Console.WriteLine("Test QBE on class with member re-initialization in ancestor constructor");
                    ListResult(result);
                }
                catch (Db4oException ex)
                {
                    System.Console.WriteLine("Db4o Exception: " + ex.Message);
                }
                catch (Exception ex)
                {
                    System.Console.WriteLine("System Exception: " + ex.Message);
                }
                finally
                {
                    CloseDatabase();
                }
            }
        }

        // end Test2

        private static void Test3()
        {
            IObjectContainer container = Database();
            if (container != null)
            {
                try
                {
                    // Pilot2 uses static value to initializ points 
                    Pilot2 pilot = new Pilot2("Kimi Raikonnen", 100);
                    container.Set(pilot);
                    container.Commit();
                    // QBE returns result with wrong points
                    IObjectSet result = container.Get(new Pilot2("Kimi Raikonnen", 100));
                    System.Console.WriteLine("Test QBE on class with member initialization using static value");
                    ListResult(result);
                }
                catch (Db4oException ex)
                {
                    System.Console.WriteLine("Db4o Exception: " + ex.Message);
                }
                catch (Exception ex)
                {
                    System.Console.WriteLine("System Exception: " + ex.Message);
                }
                finally
                {
                    CloseDatabase();
                }
            }
        }

        // end Test3

        private static void Test4()
        {
            IObjectContainer container = Database();
            if (container != null)
            {
                try
                {
                    // Pilot2Derived is derived from class with initialization of points member using static value
                    Pilot2Derived pilot = new Pilot2Derived("Kimi Raikonnen", 100);
                    container.Set(pilot);
                    container.Commit();
                    // QBE returns result with wrong points
                    IObjectSet result = container.Get(new Pilot2Derived("Kimi Raikonnen", 100));
                    System.Console.WriteLine("Test QBE on class derived from a class with member initialization using static member");
                    ListResult(result);
                }
                catch (Db4oException ex)
                {
                    System.Console.WriteLine("Db4o Exception: " + ex.Message);
                }
                catch (Exception ex)
                {
                    System.Console.WriteLine("System Exception: " + ex.Message);
                }
                finally
                {
                    CloseDatabase();
                }
            }
        }

        // end Test4

        private static void ListResult(IObjectSet result)
        {
            System.Console.WriteLine(result.Count);
            foreach (object obj in result)
            {
                System.Console.WriteLine(obj);
            }
        }

        // end ListResult
    }
}

File: reference/reference.vb.net/Db4objects.Db4odoc/QBE/QBEExample.vb
' Copyright (C) 2007 db4objects Inc. http://www.db4o.com

Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.IO

Imports Db4objects.Db4o
Imports Db4objects.Db4o.Ext
Imports Db4objects.Db4o.Query

Namespace Db4objects.Db4odoc.QBE
    Class QBEExample
        Private Const Db4oFileName As String = "reference.db4o"

        Private Const ObjectCount As Integer = 10

        Private Shared _container As IObjectContainer = Nothing

        Public Shared Sub Main(ByVal args As String())
            File.Delete(Db4oFileName)

            Test()
            Test1()
            Test2()
            Test3()
            Test4()
        End Sub
        ' end Main

        Private Shared Function Database() As IObjectContainer
            If _container Is Nothing Then
                Try
                    _container = Db4oFactory.OpenFile(Db4oFileName)
                Catch ex As DatabaseFileLockedException
                    System.Console.WriteLine(ex.Message)
                End Try
            End If
            Return _container
        End Function

        ' end Database

        Private Shared Sub CloseDatabase()
            If _container IsNot Nothing Then
                _container.Close()
                _container = Nothing
            End If
        End Sub

        ' end CloseDatabase

        Private Shared Sub Test()
            Dim container As IObjectContainer = Database()
            If container IsNot Nothing Then
                Try
                    Dim pilot As New Pilot("Kimi Raikonnen", 100)
                    container.[Set](pilot)
                    container.Commit()
                    Dim result As IObjectSet = container.[Get](New Pilot("Kimi Raikonnen", 100))
                    ListResult(result)
                Catch ex As Db4oException
                    System.Console.WriteLine("Db4o Exception: " + ex.Message)
                Catch ex As Exception
                    System.Console.WriteLine("System Exception: " + ex.Message)
                Finally
                    CloseDatabase()
                End Try
            End If
        End Sub

        ' end Test

        Private Shared Sub Test1()
            Dim container As IObjectContainer = Database()
            If container IsNot Nothing Then
                Try
                    ' Pilot1 contains re-initialisation in the constructor
                    Dim pilot As New Pilot1("Kimi Raikonnen", 100)
                    container.[Set](pilot)
                    container.Commit()
                    ' QBE returns result with wrong points
                    Dim result As IObjectSet = container.[Get](New Pilot1("Kimi Raikonnen", 100))
                    System.Console.WriteLine("Test QBE on class with member re-initialization in constructor")
                    ListResult(result)
                Catch ex As Db4oException
                    System.Console.WriteLine("Db4o Exception: " + ex.Message)
                Catch ex As Exception
                    System.Console.WriteLine("System Exception: " + ex.Message)
                Finally
                    CloseDatabase()
                End Try
            End If
        End Sub

        ' end Test1

        Private Shared Sub Test2()
            Dim container As IObjectContainer = Database()
            If container IsNot Nothing Then
                Try
                    ' Pilot1Derived derives the constructor with re-initialisation
                    Dim pilot As New Pilot1Derived("Kimi Raikonnen", 100)
                    container.[Set](pilot)
                    container.Commit()
                    ' QBE returns result with wrong points
                    Dim result As IObjectSet = container.[Get](New Pilot1Derived("Kimi Raikonnen", 100))
                    System.Console.WriteLine("Test QBE on class with member re-initialization in ancestor constructor")
                    ListResult(result)
                Catch ex As Db4oException
                    System.Console.WriteLine("Db4o Exception: " + ex.Message)
                Catch ex As Exception
                    System.Console.WriteLine("System Exception: " + ex.Message)
                Finally
                    CloseDatabase()
                End Try
            End If
        End Sub

        ' end Test2

        Private Shared Sub Test3()
            Dim container As IObjectContainer = Database()
            If container IsNot Nothing Then
                Try
                    ' Pilot2 uses static value to initializ points
                    Dim pilot As New Pilot2("Kimi Raikonnen", 100)
                    container.[Set](pilot)
                    container.Commit()
                    ' QBE returns result with wrong points
                    Dim result As IObjectSet = container.[Get](New Pilot2("Kimi Raikonnen", 100))
                    System.Console.WriteLine("Test QBE on class with member initialization using static value")
                    ListResult(result)
                Catch ex As Db4oException
                    System.Console.WriteLine("Db4o Exception: " + ex.Message)
                Catch ex As Exception
                    System.Console.WriteLine("System Exception: " + ex.Message)
                Finally
                    CloseDatabase()
                End Try
            End If
        End Sub

        ' end Test3

        Private Shared Sub Test4()
            Dim container As IObjectContainer = Database()
            If container IsNot Nothing Then
                Try
                    ' Pilot2Derived is derived from class with initialization of points member using static value
                    Dim pilot As New Pilot2Derived("Kimi Raikonnen", 100)
                    container.[Set](pilot)
                    container.Commit()
                    ' QBE returns result with wrong points
                    Dim result As IObjectSet = container.[Get](New Pilot2Derived("Kimi Raikonnen", 100))
                    System.Console.WriteLine("Test QBE on class derived from a class with member initialization using static member")
                    ListResult(result)
                Catch ex As Db4oException
                    System.Console.WriteLine("Db4o Exception: " + ex.Message)
                Catch ex As Exception
                    System.Console.WriteLine("System Exception: " + ex.Message)
                Finally
                    CloseDatabase()
                End Try
            End If
        End Sub

        ' end Test4

        Private Shared Sub ListResult(ByVal result As IObjectSet)
            System.Console.WriteLine(result.Count)
            Dim obj As Object
            For Each obj In result
                System.Console.WriteLine(obj)
            Next
        End Sub

        ' end ListResult
    End Class
End Namespace